README.md

```diff
@@ -364,6 +366,40 @@ If a `.spec.type` is set, the `type` entry in the binding `Secret` **MUST** be s
 
 If the modification of the Application resource is completed successfully, the `Ready` condition status **MUST** be set to `True`.  If the modification of the Application resource is not completed successfully the `Ready` condition status **MUST NOT** be set to `True`.
 
+## Reconciliation Duties
+
+There are scenarios where the reconciler that processes the bindings of a provisioned service is different than the reconciler that will project those bindings into the Application. In such cases the `ServiceBinding` CR author **MUST** set `.spec.application.autoProjection` to `false`, signalling the separation of reconciliation duties between `.spec.service` and `.spec.application`.  The default value of this property is `true`.  
+
+When `.spec.application.autoProjection` is set to `false`, the reconciler that processes `.spec.service` **MUST** reflect the generated secret, which is ready for Application projection, as `.status.binding.name`.  The reconciler that processes `.spec.application` can therefore watch for this particular `.status` state, and then proceed with Application projection.  When projection is completed the latter reconciler is responsible for ensuring the appropriate `.status.condition` states are set, such as the `Ready` condition.
```

Comment:
Two reconcilers both handing the same resource is generally an anti-pattern that can led to races and thrashing the API Server with mutations.

![giphy](https://user-images.githubusercontent.com/302992/87332694-f68c7880-c509-11ea-9acc-2234e0697241.gif)

There is also state that can only be defined by the former reconciler. For example, if a mapping template is invalid, or if there was any other error generating the secret, that error needs to be reflected.

Reply:
Ya - I thought about introducing a new Condition of type `SecretReady`, that the former reconciler would use to indicate failure and that the latter reconciler could use to abort waiting for `.status.binding`.  Would that address your concern?

Reply:
it's `duck type` season 😆

Reply:
> I thought about introducing a new Condition of type SecretReady, that the former reconciler would use to indicate failure and that the latter reconciler could use to abort waiting for .status.binding. Would that address your concern?

Partially.

Adding a new condition would give a way to report incremental status for aspects of the reconciliation. This wouldn't solve the issue where there are two reconcilers actively trying to write to the same resource. While possible, this behavior is tricky to get right and requires cooperation on both ends. Given that one of those ends is open to extension, I'm not optimistic about the ecosystem being well behaved.

While less informative to watchers of the `ServiceBinding` resource, I'd rather see a single reconciler be responsible for writing the status. When that reconciler is not responsible for projecting the binding into the application, the `ServiceBinding`'s status likewise is not responsible for reporting the status of work it's not responsible for.

Reply:
> the `ServiceBinding`'s status likewise is not responsible for reporting the status of work it's not responsible for

I would like to break the assumption that a `ServiceBinding` resource is directly tied to a single processing unit.  I have always viewed `ServiceBinding` and its status as the encapsulation of a set of binding activities:
* processing the binding secret of the provisioned service
* creating mappings
* projecting the final bindings (mount and env vars)

The `ServiceBinding` resource should be agnostic to the fact that we may have 1, 2 or 3 different processors doing the work - and given that this work is sequential (never in parallel), I don't think there would be a race condition for carrying out the work and updating the different Condition types.  

Essentially, by defining the status Condition for each stage we then have well defined hand-off points of control.

Reply:
> I would like to break the assumption that a ServiceBinding resource is directly tied to a single processing unit. I have always viewed ServiceBinding and its status as the encapsulation of a set of binding activities

My issue is that this model goes against the current best practice to have a single reconciler for a resource. It's certainly possible to have multiple reconcilers for a single resource, but things can quickly go sideways unless every component trying to manage that resource has a clear understanding of its boundaries. But we're talking about this as an extension point, so I don't trust those external reconcilers to be well behaved. When they inevitably do misbehave, users will blame the `ServiceBinding` since that's what they created and all they can see.

An alternative approach that would be robust is to create an internal SPI resource that is created by the ServiceBinding reconciler representing the extensible sub-task. A third-party operator can manage the reconciliation of that resource, and the ServiceBinding can watch for status changes on the internal resource and report that status back out in a meaningful way. There is precedent for this model in Knative with its Ingress (ingresses.networking.internal.knative.dev) resource.  The KIngress is created by the Knative Serving controller, but the reconciler for KIngress is provided by an independent networking plugin.

Reply:
that's a good suggestion @scothis 

I wonder though:  if we have a corresponding status Condition for each of the delineated stages of `ServiceBinding`, and this is defined at the spec level, isn't that enough to provide the clear understanding of the boundaries?    

I would think that setting a particular status Condition, where you "own" that status Condition on the shared resource, would be easier to implement than creating a new internal resource + manage its lifecycle + setting its status Condition.

Reply:
> I wonder though: if we have a corresponding status Condition for each of the delineated stages of `ServiceBinding`, and this is defined at the spec level, isn't that enough to provide the clear understanding of the boundaries?

I don't think this behavior is good idea and don't want any suggestions or nudges towards that behavior in the core spec because it's too easy for one poor third-party implementation to cause a whole cluster to go sideways. Individual implementations are of course free to do what they see fit for their own needs by going beyond the spec.

> I would think that setting a particular status Condition, where you "own" that status Condition on the shared resource, would be easier to implement than creating a new internal resource + manage its lifecycle + setting its status Condition.

This is a case where the implementation that seems easier leads to instability. It'll ultimately be easier for everyone to do the right thing from the beginning.